Source/WebKit/WebProcess/WebPage/gtk/AcceleratedSurfaceDMABuf.h

```diff
@@ -71,88 +72,102 @@ class AcceleratedSurfaceDMABuf final : public AcceleratedSurface, public IPC::Me
         ~RenderTargetEGLImage();
 
     private:
-        void didRenderFrame() override;
-        void didDisplayFrame() override;
-
-        EGLImage m_backImage { nullptr };
-        EGLImage m_frontImage { nullptr };
-        EGLImage m_displayImage { nullptr };
+        EGLImage m_image { nullptr };
     };
 #endif
 
     class RenderTargetSHMImage final : public RenderTargetColorBuffer {
     public:
         static std::unique_ptr<RenderTarget> create(uint64_t, const WebCore::IntSize&);
-        RenderTargetSHMImage(uint64_t, const WebCore::IntSize&, Ref<ShareableBitmap>&&, ShareableBitmapHandle&&, Ref<ShareableBitmap>&&, ShareableBitmapHandle&&, Ref<ShareableBitmap>&&, ShareableBitmapHandle&&);
+        RenderTargetSHMImage(uint64_t, const WebCore::IntSize&, Ref<ShareableBitmap>&&, ShareableBitmapHandle&&);
         ~RenderTargetSHMImage() = default;
 
     private:
         void didRenderFrame() override;
-        void didDisplayFrame() override;
 
-        Ref<ShareableBitmap> m_backBitmap;
-        Ref<ShareableBitmap> m_frontBitmap;
-        Ref<ShareableBitmap> m_displayBitmap;
+        Ref<ShareableBitmap> m_bitmap;
     };
 
     class RenderTargetTexture final : public RenderTarget {
     public:
         static std::unique_ptr<RenderTarget> create(uint64_t, const WebCore::IntSize&);
-        RenderTargetTexture(uint64_t, const WebCore::IntSize&, WTF::UnixFileDescriptor&&, unsigned, WTF::UnixFileDescriptor&&, unsigned, WTF::UnixFileDescriptor&&, unsigned, uint32_t format, uint32_t offset, uint32_t stride, uint64_t modifier);
+        RenderTargetTexture(uint64_t, const WebCore::IntSize&, WTF::UnixFileDescriptor&&, unsigned, uint32_t format, uint32_t offset, uint32_t stride, uint64_t modifier);
         ~RenderTargetTexture();
 
     private:
         void willRenderFrame() const override;
-        void didRenderFrame() override;
-        void didDisplayFrame() override;
 
-        unsigned m_backTexture { 0 };
-        unsigned m_frontTexture { 0 };
-        unsigned m_displayTexture { 0 };
+        unsigned m_texture { 0 };
+    };
+
+    class SwapChain {
+        WTF_MAKE_NONCOPYABLE(SwapChain);
+    public:
+        explicit SwapChain(uint64_t);
+        ~SwapChain() = default;
+
+        void resize(const WebCore::IntSize&);
+        RenderTarget* nextTarget();
+        void releaseTarget(uint64_t);
+        void reset();
+
+        unsigned size() const { return m_freeTargets.size() + m_lockedTargets.size(); }
+
+    private:
+        enum class Type { Invalid, EGLImage, SharedMemory, Texture };
+
+        std::unique_ptr<RenderTarget> createTarget() const;
+
+        uint64_t m_surfaceID { 0 };
+        Type m_type { Type::Invalid };
+        WebCore::IntSize m_size;
+        Vector<std::unique_ptr<RenderTarget>, 3> m_freeTargets;
+        Vector<std::unique_ptr<RenderTarget>, 3> m_lockedTargets;
```

Comment:
Could we add some value with the max amount of buffers, I think there is an assert where we could use it in the code.